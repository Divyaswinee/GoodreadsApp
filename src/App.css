.app {
  background-color: #ffebcd;
  min-height: 100vh;
  display: flex;
  flex: 1;
  flex-direction: column;
}

.App-logo {
  height: 20vmin;
}

.App-link {
  color: #61dafb;
}

.searchHeader {
  display: flex;
  background-color: #d2b48c;
  flex: 0.1;
  min-height: 5em;
  max-height: 5em;
  align-items: center;
  justify-content: center;
}

.searchBody {
  display: flex;
  flex: 0.9;
  height: 100%
}

.iconDiv {
  flex: 1;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
}

.iconStyle {
  width: 20em;
  height: 20em;
  border-radius: 50%;
}

.labelStyle {
  margin-top: 0.5em;
  font-size: 5rem;
  color: #8b4513;
  font-family: "Times New Roman", Times, serif;
}
