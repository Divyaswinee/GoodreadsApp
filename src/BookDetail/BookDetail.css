.containerBody {
	display: flex;
	flex: 1;
	flex-direction: row;
}

.containerLeft {
	display: flex;
	flex: 0.4;
}

.containerRight {
	display: flex;
	flex: 0.6;
	flex-direction: column;
	font-family: "Times New Roman", Times, serif;
	color: #8b4513;
	padding: 2em;
}

.imgStyle {
	flex: 1;
	margin: 2em;
	border-color: transparent;
}

.titleDiv {
	flex: 0.2;
	font-size: 3rem;
	font-weight: bold;
}

.authorDiv {
	flex: 0.1;
	font-size: 2rem;
}

.ratingDiv {
	flex: 0.05;
	display: flex;
	align-items: center;
	margin-top: 1em;
}

.descriptionDiv {
	display: block;
	margin-top: 3em;
}

.starRatingStyles {
	padding-right: 1em;
}

