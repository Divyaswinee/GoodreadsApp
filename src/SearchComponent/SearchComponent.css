.searchContainer {
	display: flex;
	flex: 1;
	width: 100%;
	justify-content: center;
}

.searchBoxCustomize {
	width: 50em;
	border-radius: 25px;
}
