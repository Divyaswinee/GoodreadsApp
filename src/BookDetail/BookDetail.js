import React, { Component } from "react";
import axios from 'axios';
import proxify from 'proxify-url';
import StarRatingComponent from 'react-star-rating-component';
import './BookDetail.css';

const API_URL = 'https://www.goodreads.com/book/show/';
const API_KEY = 'vZt6PSQ52iaN3bFKvEwOlQ';

class BookDetail extends Component {
	state = {
    bookDescription: ''
  };

  componentDidMount() {
  	this.getBookDescription(this.props.bookInfo);
  }

  componentWillReceiveProps(nextProps) {
  	const { bookInfo } = this.props;
  	if(nextProps.bookInfo.label !== bookInfo.label) {
  		this.getBookDescription(nextProps.bookInfo);
  	}
  }

  getBookDescription = (book) => {
  	const bookID = book.id;
  	const url = `${API_URL}${bookID}?key=${API_KEY}`;
    const proxyUrl = proxify(url, { inputFormat: 'xml' });
    axios.get(proxyUrl)
    	.then(({ data }) => {
        return data;
      })
      .then(dataNext => {
        const bookResult = dataNext.query.results.GoodreadsResponse.book;
        this.setState({
          bookDescription: bookResult.description
        });
      })
      .catch(error => {
        console.log('Error', error.toString());
      });
  }

  render() {
  	const { bookInfo } = this.props;
  	return (
  		<div className='containerBody'>
  			<div className='containerLeft'>
  				<img src={bookInfo.URL} alt='' className='imgStyle'/>
  			</div>
  			<div className='containerRight'>
  				<div className='titleDiv'>
  					<span>{bookInfo.bookName}</span>
  				</div>
  				<div className='authorDiv'>
  					<span>by {bookInfo.author}</span>
  				</div>
  				<div className='ratingDiv'>
  					<StarRatingComponent 
		          name="rating" 
		          starCount={5}
		          value={bookInfo.averageRating}
		          className='starRatingStyles'
		          starColor={'#ff8c00'}
		          emptyStarColor={'#ffffff'}
		        />
  					<span>{bookInfo.averageRating}</span>
  				</div>
  				{this.state.bookDescription !== 'Undefined' && <div className='descriptionDiv'>
  					{this.state.bookDescription}
  				</div>}
  			</div>
  		</div>
  	);
  }

}

export default BookDetail;