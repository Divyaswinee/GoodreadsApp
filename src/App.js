import React, { Component } from 'react';
import SearchComponent from './SearchComponent/SearchComponent.js';
import BookDetail from './BookDetail/BookDetail.js';
import './App.css';

class App extends Component {
  state = {
    selectedBook: {},
    showBookDetail: false
  }
  getSearchBookData = (selectedBook) => {
    this.setState({
      selectedBook,
      showBookDetail: true 
    });
  }
  showDetail = (value) => {
    this.setState({
      showBookDetail: value 
    });
  }
  render() {
    return (
      <div className="app">
        <div className="searchHeader">
          <SearchComponent onSearchCallback={this.getSearchBookData} showDetailCallBack={this.showDetail} />
        </div>
        <div className="searchBody">
          {this.state.showBookDetail && <BookDetail bookInfo={this.state.selectedBook}/>}
          {!this.state.showBookDetail &&
            <div className='iconDiv'>
              <img src='./books.ico' alt='' className='iconStyle'/>
              <span className='labelStyle'>BOOKLET</span>
            </div>
          }
        </div>
      </div>
    );
  }
}

export default App;
