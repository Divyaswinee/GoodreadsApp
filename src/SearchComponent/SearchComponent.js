import React, { Component } from "react";
import AsyncSelect from 'react-select/lib/Async';
import axios from 'axios';
import proxify from 'proxify-url';
import PropTypes from 'prop-types';
import './SearchComponent.css';

const API_URL = 'https://www.goodreads.com/search/index.xml';
const API_KEY = 'vZt6PSQ52iaN3bFKvEwOlQ';

class SearchComponent extends Component {
	state = {
    searchText: ''
  };

  static propTypes = {
    onSearchCallback: PropTypes.func
  }

  getSearchResult = (searchTerm) => {
    const url = `${API_URL}?key=${API_KEY}&q=${searchTerm}&search=title`;
    const proxyUrl = proxify(url, { inputFormat: 'xml' });
    return new Promise(resolve => {
      axios.get(proxyUrl)
        .then(({ data }) => {
          return data;
        })
        .then(datatwo => {
          console.log('datatwo ', datatwo);
          const searchResult = datatwo.query.results.GoodreadsResponse.search;
          const allResult = searchResult.results;
          const workResult = allResult && allResult.work && allResult.work.length > 0 && allResult.work.map((current) => {
            return {
              value: current['id']['content'],
              author: current['best_book']['author']['name'],
              label: current['best_book']['title'],
              bookName: current['best_book']['title'],
              URL: current['best_book']['image_url'],
              averageRating: current['average_rating'],
              id: current['id']['content']
            }
          })
          console.log('workResult ', workResult);
          resolve(workResult);
        })
        .catch(error => {
          console.log('Error', error.toString());
        });
    });
  };

  searchHandleChange = (value) => {
    this.setState({ searchText: value });
    if(value.length === 0) {
      return this.props.showDetailCallBack(false);
    }
    this.props.onSearchCallback(value);
  };

  render() {
  	return (
  		<div className="searchContainer">
  	  	<AsyncSelect
          className="searchBoxCustomize"
          value={this.state.searchText}
          onChange={this.searchHandleChange}
          placeholder="Search Book"
          loadOptions={this.getSearchResult.bind(this)}
          defaultOptions
          backspaceRemoves={true}
          cache={false}
        />
      </div>
  	);
  }
}

export default SearchComponent;