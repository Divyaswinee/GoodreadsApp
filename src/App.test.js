import React from 'react';
import ReactDOM from 'react-dom';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from './App';
import SearchComponent from './SearchComponent/SearchComponent.js';
import BookDetail from './BookDetail/BookDetail.js';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
	let wrapper;
	wrapper = mount(<App />);

	const component = shallow(<App />);
  it ('component renders correctly', () => {
    expect(component).toMatchSnapshot();
  });

	it('renders without crashing', () => {
	  const div = document.createElement('div');
	  ReactDOM.render(<App />, div);
	  ReactDOM.unmountComponentAtNode(div);
	});

	it ('check child component SearchComponent Present or not', () => {
    expect(wrapper.find(SearchComponent).length).toEqual(1);
  });
  
  it ('check child component BookDetail Present or not', () => {
  	expect(wrapper.find(BookDetail).length).toEqual(0);
  	wrapper.setState({ showBookDetail: true })
    expect(wrapper.find(BookDetail).length).toEqual(1);
  });
});
