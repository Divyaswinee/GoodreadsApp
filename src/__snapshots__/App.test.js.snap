// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="searchHeader"
        >
          <SearchComponent
            onSearchCallback={[Function]}
            showDetailCallBack={[Function]}
          />
        </div>,
        <div
          className="searchBody"
        >
          <div
            className="iconDiv"
          >
            <img
              alt=""
              className="iconStyle"
              src="./books.ico"
            />
            <span
              className="labelStyle"
            >
              BOOKLET
            </span>
          </div>
        </div>,
      ],
      "className": "app",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <SearchComponent
            onSearchCallback={[Function]}
            showDetailCallBack={[Function]}
          />,
          "className": "searchHeader",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onSearchCallback": [Function],
            "showDetailCallBack": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            false,
            <div
              className="iconDiv"
            >
              <img
                alt=""
                className="iconStyle"
                src="./books.ico"
              />
              <span
                className="labelStyle"
              >
                BOOKLET
              </span>
            </div>,
          ],
          "className": "searchBody",
        },
        "ref": null,
        "rendered": Array [
          false,
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <img
                  alt=""
                  className="iconStyle"
                  src="./books.ico"
                />,
                <span
                  className="labelStyle"
                >
                  BOOKLET
                </span>,
              ],
              "className": "iconDiv",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "alt": "",
                  "className": "iconStyle",
                  "src": "./books.ico",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "BOOKLET",
                  "className": "labelStyle",
                },
                "ref": null,
                "rendered": "BOOKLET",
                "type": "span",
              },
            ],
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="searchHeader"
          >
            <SearchComponent
              onSearchCallback={[Function]}
              showDetailCallBack={[Function]}
            />
          </div>,
          <div
            className="searchBody"
          >
            <div
              className="iconDiv"
            >
              <img
                alt=""
                className="iconStyle"
                src="./books.ico"
              />
              <span
                className="labelStyle"
              >
                BOOKLET
              </span>
            </div>
          </div>,
        ],
        "className": "app",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <SearchComponent
              onSearchCallback={[Function]}
              showDetailCallBack={[Function]}
            />,
            "className": "searchHeader",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "onSearchCallback": [Function],
              "showDetailCallBack": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              false,
              <div
                className="iconDiv"
              >
                <img
                  alt=""
                  className="iconStyle"
                  src="./books.ico"
                />
                <span
                  className="labelStyle"
                >
                  BOOKLET
                </span>
              </div>,
            ],
            "className": "searchBody",
          },
          "ref": null,
          "rendered": Array [
            false,
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <img
                    alt=""
                    className="iconStyle"
                    src="./books.ico"
                  />,
                  <span
                    className="labelStyle"
                  >
                    BOOKLET
                  </span>,
                ],
                "className": "iconDiv",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "alt": "",
                    "className": "iconStyle",
                    "src": "./books.ico",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "img",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "BOOKLET",
                    "className": "labelStyle",
                  },
                  "ref": null,
                  "rendered": "BOOKLET",
                  "type": "span",
                },
              ],
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
